### PR TITLE
Creating Math service, integrating redux, add history like formulas-l…

### PR DESCRIPTION
**Pull request by dm-rybalchenko**
…ist using react context.

### CODE REVIEW

File: src/lib/editableGrid/Cell.tsx [hunk 1]
```diff
@@ -63,6 +64,15 @@ export default function Cell({
     setValue({ value: e.target.value, update: false });
   }
 
+  function onBlur(e) {
```
Comment (cybermerlin) [inline src/lib/editableGrid/Cell.tsx:73]: pls, add a comment 'for what this function need'
eg:
```
/**
 * This function need foractualize...
 */
```

File: src/plugins/calculationTry1/MathService.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import React, { createContext } from "react";
+
+
+export type THistoryFormulas = {
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/MathService.tsx]: actually no need do this like an object when it has only one property.
Or convert it to an interface and add couple of methods (add(), del(), get(), toString()|Symbol.toPrimitive) as well.

File: src/plugins/calculationTry1/MathService.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+const ERROR_MESSAGES = Object.freeze({
+  formula: "Проверьте формулу, она некорректна",
+  symbols:
+    "Некорректные символы в формуле!\n Допустимы цифры 0-9, -, +, *, /, (, ), . \n Формула должна начинаться с '='",
+});
+let historyFormulas = { history: "" };
+const MathServiceContext: React.Context<THistoryFormulas> = createContext(null);
+
+function MathService(props: { children: JSX.Element }) {
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/MathService.tsx]: when it as a HTML-like component, probably better MathServiceComponent.

File: src/plugins/calculationTry1/MathService.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  return <MathServiceContext.Provider value={historyFormulas}>{props.children}</MathServiceContext.Provider>;
+}
+
+function MathCounter(formula: string): number | Error {
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/MathService.tsx]: Counter - usually used exactly for counter (as i++)

I goes a better name MathCacl|MathCalculator|...evaluator....

File: src/plugins/calculationTry1/handlersCountCellsData.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import { MathCounter } from './MathService';
+import { storeMath, add_formula } from './storeMath';
+
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/handlersCountCellsData.ts]: no need empty row in this case

File: src/plugins/calculationTry1/handlersCountCellsData.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import { MathCounter } from './MathService';
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/handlersCountCellsData.ts]: before this line u can add an empty line (to divide libraries and project things.

File: src/plugins/calculationTry1/countCellsData.tsx [hunk 7]
```diff
@@ -1,6 +1,18 @@
 import React from "react";
-import handleKeyPress from "./handleKeyPress";
+import { handleKeyPress, handleFocus, handleBlur } from "./handlersCountCellsData";
+import HistoryFormulas from "./historyFormulas";
+import { MathService } from "./MathService";
 
-export default function CountCellsData(prop: JSX.Element) {
-  return <div onKeyPress={handleKeyPress}>{prop}</div>;
+
+export default function CountCellsData(props: { children: JSX.Element }) {
```
Comment (cybermerlin) [inline src/plugins/calculationTry1/countCellsData.tsx]: CalculateCellsForumals - probably

**Review by cybermerlin — CHANGES_REQUESTED** [review]

File: src/plugins/calculationTry1/MathService.tsx (hunk 2 above)
Comment (dm-rybalchenko) [reply src/plugins/calculationTry1/MathService.tsx]: Discussed about React-comparing changes variables in React.Context.

File: src/lib/editableGrid/Table.tsx [hunk 8]
```diff
@@ -74,14 +74,18 @@ export default function Table({
...
     ({ index, style }) => {
       const row = rows[index];
+
       prepareRow(row);
+
       return (
         <div {...row.getRowProps({ style })} className="tr">
           {row.cells.map(cell => (
-            <div {...cell.getCellProps()} className="td">
+            <div {...cell.getCellProps()} tabIndex={tabidexCell++} className="td">
```
Comment (cybermerlin) [inline src/lib/editableGrid/Table.tsx]: tabindex - no camelCase for standard attrs

File: src/lib/editableGrid/Cell.tsx [hunk 9]
```diff
@@ -63,6 +64,18 @@ export default function Cell({
     setValue({ value: e.target.value, update: false });
   }
 
+  /**
+ * This function need for preventing saving unended fomulas (saves last etered formula in a cell)
```
Comment (cybermerlin) [inline src/lib/editableGrid/Cell.tsx]: orfo pls

File: src/lib/editableGrid/Table.tsx (hunk 8 above)
Comment (dm-rybalchenko) [reply src/lib/editableGrid/Table.tsx]: Discussed property in index.d.ts of React.

**Comment by cybermerlin** [comment]
tests was fallen

File: src/plugins/math/pureMath/MathCalculator.test.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  it('Returns an Error for uncorrect syntax in a formula', () => {
+    const SOLUTION = MathCalculator('=2+2(3+3)))');
+
+    expect(typeof SOLUTION).toEqual(typeof new Error());
+  });
+
+  it('Returns an Error when solution of a formula is Infinity', () => {
+    const SOLUTION = MathCalculator('=100/0');
+
+    expect(typeof SOLUTION).toEqual(typeof new Error());
```
Comment (cybermerlin) [inline src/plugins/math/pureMath/MathCalculator.test.ts]: `expect..toBeInstanceOf(Error)` ?

File: src/plugins/math/pureMath/MathService.tsx [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import React, { createContext } from 'react';
```
Comment (cybermerlin) [inline src/plugins/math/pureMath/MathService.tsx]: no need import React

File: src/plugins/math/pureMath/handlersCountCellsData.test.tsx [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+});
+
+afterEach(() => {
+  if (container instanceof HTMLDivElement) {
+    act(() => root.unmount());
+    container.remove();
+    container = null;
+  }
+});
+
```
Comment (cybermerlin) [inline src/plugins/math/pureMath/handlersCountCellsData.test.tsx]: add one Enter

File: src/plugins/math/pureMath/handlersCountCellsData.test.tsx [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+      root.render(
+        <div tabIndex={33} onKeyPress={handleKeyPress} onFocus={handleFocus}>
+          <div tabIndex={44} className="data-input-number"></div>
+        </div>
+      );
+    });
+
+    let cell = document.querySelector('.data-input-number') as HTMLElement;
+
+    expect(cell).not.toBeNull();
```
Comment (cybermerlin) [inline src/plugins/math/pureMath/handlersCountCellsData.test.tsx:36]: add a line after this line

Comment (cybermerlin) [inline src/plugins/math/pureMath/handlersCountCellsData.test.tsx]: after casting into an HTMLElement you don't have any rights to check nullable values.

File: src/plugins/math/pureMath/storeMath.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import { AnyAction } from '@reduxjs/toolkit';
+import { legacy_createStore as createStore } from 'redux';
+
+
```
Comment (cybermerlin) [inline src/plugins/math/pureMath/storeMath.ts:4]: look at the conversation of meeting
no need switch with one case
camelCase for functions naming
why the type  named through the constant?

**Review by cybermerlin — APPROVED** [review]

**Pull request merged**